File: backend/internal/model/post.go
package model

import (
	"time"
)

type Post struct {
	ID int64 `json:"id"`

	UserAccountID int64 `json:"user_account_id"`
	ImageID       int64 `json:"image_id"`

	Title   string `json:"title"`
	Content string `json:"content"`
	Likes   int    `json:"likes"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

func (p *Post) FromRow(row Scannable) error {
	return row.Scan(
		&p.ID,
		&p.UserAccountID,
		&p.ImageID,
		&p.Title,
		&p.Content,
		&p.Likes,
		&p.CreatedAt,
		&p.UpdatedAt,
	)
}

type APIPost struct {
	ID int64 `json:"id"`

	UserAccountID int64 `json:"user_account_id"`
	ImageID       int64 `json:"image_id"`

	Title   string `json:"title"`
	Content string `json:"content"`
	Likes   int    `json:"likes"`

	Annotations []*Annotation `json:"annotations"`
	Tags        []*Tag        `json:"tags"`

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

func (p *Post) ToAPIPost(annotations []*Annotation, tags []*Tag) *APIPost {
	if annotations == nil {
		annotations = []*Annotation{}
	}
	if tags == nil {
		tags = []*Tag{}
	}
	return &APIPost{
		ID:            p.ID,
		UserAccountID: p.UserAccountID,
		ImageID:       p.ImageID,
		Title:         p.Title,
		Content:       p.Content,
		Likes:         p.Likes,
		Annotations:   annotations,
		Tags:          tags,
		CreatedAt:     p.CreatedAt,
		UpdatedAt:     p.UpdatedAt,
	}
}

File: frontend/src/types/api/post.ts
import { z } from "zod";
import { AnnotationSchema } from "../db/annotation";
import { TagSchema } from "../db/tag";
import { UserAccountSchema } from "../db/user";

const API_PostSchema = z.object({
    id: z.number(),

    image_id: z.number(),
    title: z.string(),
    content: z.string(),
    created_at: z.string().datetime(),
    updated_at: z.string().datetime(),

    user_account: UserAccountSchema,
    annotations: z.array(AnnotationSchema),
    tags: z.array(TagSchema),
})

type API_Post = z.infer<typeof API_PostSchema>;

export { API_PostSchema };
export type { API_Post }; 